R/gitlab.R

```diff
@@ -0,0 +1,72 @@
+isGitlabURL <- function(url) {
+  is.string(url) && grepl("^http(?:s)?://(?:www|api).gitlab.(org|com)", url, perl = TRUE)
+}
+
+canUseGitlabDownloader <- function() {
+  all(packageVersionInstalled(devtools = "1.9.1", httr = "1.0.0"))
+}
+
+gitlabDownload <- function(url, destfile, ...) {
+  onError(1,{
+    gitlab_user    <- gitlab_user
+    gitlab_pwd     <- gitlab_pwd
+    authenticate   <- yoink("httr", "authenticate")
+    GET            <- yoink("httr", "GET")
+    content        <- yoink("httr", "content")
+
+    user <- gitlab_user(quiet = TRUE)
+    pwd <- gitlab_pwd(quiet = TRUE)
+    auth <- if(!is.null(user) & !is.null(pwd)) {
+      authenticate(user, pwd, type = "basic")
+    } else {
+      list()
+    }
+
+    request <- GET(url, auth)
+    if(request$status == 401) {
+      warning("Failed to download package from gitlab: not authorized. ",
+              "Did you set GITLAB_USERNAME and GITLAB_PASSWORD env vars?",
+              call. = FALSE)
+      return(1)
+    }
+    writeBin(content(request, "raw"), destfile)
```

Comment:
Should we double-check that we got a 200 response code before writing the request to disk?